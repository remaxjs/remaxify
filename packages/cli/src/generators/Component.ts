import * as t from '@babel/types';
import { addNamed } from '@babel/helper-module-imports';
import generator from '@babel/generator';
import traverse from '@babel/traverse';
import prettier from 'prettier';
import * as sander from 'sander';
import * as path from 'path';
import { options } from '../options';
import { RUNTIME_PACKAGE } from '../constants';
import { File } from '@babel/types';
import { SourceType } from '../sourceType';

const REMAXIFY_COMPONENT = 'RemaxifyComponent';
const Component = SourceType.Component;

export default function generate(ast: File | undefined, filename: string): OutputFile {
  if (!ast) {
    throw new Error('Component 文件无法分析 ast');
  }

  filename = path.join(options.rootDir, filename);

  traverse(ast, {
    CallExpression(nodePath) {
      const { node } = nodePath;

      if (!t.isIdentifier(node.callee)) {
        return;
      }

      if (node.callee.name !== Component) {
        return;
      }

      const binding = nodePath.scope.getBinding(Component);

      if (binding) {
        return;
      }

      const appId = addNamed(nodePath, Component, RUNTIME_PACKAGE, {
        nameHint: REMAXIFY_COMPONENT,
      });
      nodePath.replaceWith(t.callExpression(appId, node.arguments));
      nodePath.stop();
    },
  });

  const code = generator(ast).code;
  const content = prettier.format(code, {
    parser: 'babel',
  });

  if (!options.dry) {
    sander.writeFile(options.output, filename, content);
  }

  return {
    filename,
    content,
  };
}
