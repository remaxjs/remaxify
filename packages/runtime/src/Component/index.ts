import * as React from 'react';

export default function Component<P = any, S = any>(
  componentConfig: { [key]: string },
  render: () => void
): React.ComponentType<P, S> {
  return class Component extends React.Component<P, S> {
    constructor(props: any) {
      super(props);

      Object.keys(componentConfig).forEach(key => {
        if (typeof componentConfig[key] === 'function') {
          (this as any)[key] = componentConfig[key].bind(this);
          return;
        }

        (this as any)[key] = componentConfig[key];
      });

      this.state = componentConfig.data;
    }

    public data = this.state;

    public setData(payload: any, callback?: () => void) {
      return this.setState(payload, callback);
    }

    public render() {
      return render.bind(this)();
    }
  };
}
