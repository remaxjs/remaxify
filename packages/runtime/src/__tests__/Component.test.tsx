import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import Component from '../Component';

describe('Component', () => {
  it('works', () => {
    const mockShow = jest.fn();
    const config: any = {
      data: {
        propA: 'A',
      },

      onShow(...params: any) {
        mockShow(...params);
      },

      onTap() {
        this.setData({
          propA: 'AA',
        });
      },
    };
    const View = function ({ onTap, children }: any) {
      return React.createElement('view', { onTap, children });
    };

    function render(this: any) {
      return (
        <View onTap={this.onTap}>
          {this.state.propA}
          {this.props.propB}
        </View>
      );
    }
    const CompComponent = Component(config, render);

    const testRenderer = TestRenderer.create(<CompComponent propB="B" />);
    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer.root.findByType('view').props.onTap();
    expect(testRenderer.toJSON()).toMatchSnapshot();

    const options = { optionA: 'A' };
    testRenderer.root.instance.onShow(options);
    expect(mockShow).toBeCalledWith(options);
  });
});
