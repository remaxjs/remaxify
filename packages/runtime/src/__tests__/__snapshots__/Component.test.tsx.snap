// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component works 1`] = `
<view
  onTap={[Function]}
>
  A
  B
</view>
`;

exports[`Component works 2`] = `
<view
  onTap={[Function]}
>
  AA
  B
</view>
`;
